the-collab-lab/tcl-36-smart-shopping-list
# User Story 09 Filter List

## Description

1. Added search input to ListView that allows users to enter characters and their list will be filtered if an item name contains those characters, included useRef focus property so that when the inventory list is rendered for the first time the input field is focused for the user. If a user clicks on the x the input field is cleared and the filter clears showing the entire list.
2. Styled the input field adding an icon to indicate it is used to search.
3. Refactored created a new component to display the list - ListLayout to centralize the logic based on the database items.
4. Refactored the success message and error message in AddItem to utilize hot toast dependency -- this fixed the previous memory leak error.
5. Added input focus via useRef hook to AddItem input field so the input is focused when the user loads the page.

## Related Issue

closes #9 

## Acceptance Criteria

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should not only match from the start of the string)


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    ✓| :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
![no input field](https://user-images.githubusercontent.com/89806097/152882093-fad287ee-8bb5-405c-845d-4b74579aa515.png)

### After

![filtering](https://user-images.githubusercontent.com/89806097/152882125-b503ff3c-c1c2-48ac-b03c-9a545045e041.png)

<img width="787" alt="afterFilter" src="https://user-images.githubusercontent.com/83471057/153452348-de5a73c6-b606-499f-b2d4-84e2abaa2f8b.png">

![toast success message](https://user-images.githubusercontent.com/89806097/152882142-e0c37db9-a95b-4ca3-9384-2d518da330cc.png)

<img width="519" alt="loading" src="https://user-images.githubusercontent.com/83471057/153453122-95d73e0f-03fd-4df7-be70-7973619288c6.png">

## Testing Steps / QA Criteria

1. From your terminal fetch branch information from the remote repository using `git fetch origin ld-am-filter-list`, and then checkout our branch with `git checkout ld-am-filter-list`, update dependencies running 'npm install' or 'npm i'.
2. Navigate to /listView, if you don't have any items add a few items to your list, enter characters to filter your list by and make sure items were filtered correctly.
3. Clear the input by clicking the 'x' button next to the input or you can use 'tab' and press 'Enter'
4. Navigate to /addItem check success and error messages for adding an item and attempting to add a duplicate item.
